### ISSUE TITLE
Tracking/chaining find clump throws C++ error

### ISSUE THREAD

**Issue by David-Baddeley** [python-microscopy/python-microscopy#49]
**[Original report](https://bitbucket.org/david_baddeley/python-microscopy/issue/49) by Bennett Rollins (Bitbucket: [BenRollins](https://bitbucket.org/BenRollins), ).**

----------------------------------------

Ran Tracking/chaining > Find clumps in VisGUI. Tried to color points by clumpIndex. Received the following error.

```
#!python

Exception occurred in traits notification handler for object: <PYME.LMVis.layers.pointcloud.PointCloudRenderLayer object at 0x1835223cb0>, trait: clim, old value: [30.0, 828.0], new value: [1.0, 8000.0]
Traceback (most recent call last):
  File "/anaconda2/lib/python2.7/site-packages/traits/trait_notifiers.py", line 519, in _dispatch_change_event
    self.dispatch( handler, *args )
  File "/anaconda2/lib/python2.7/site-packages/traits/trait_notifiers.py", line 614, in dispatch
    handler( *args )
  File "/anaconda2/lib/python2.7/site-packages/traitsui/editor.py", line 366, in _update_editor
    self.update_editor()
  File "/Users/bennettrollins/python-microscopy/PYME/ui/custom_traits_editors.py", line 131, in update_editor
    self.control.SetData(self.factory.data(), *self.value)
  File "/anaconda2/lib/python2.7/site-packages/wx-3.0-osx_cocoa/wx/_core.py", line 16711, in __getattr__
    raise PyDeadObjectError(self.attrStr % self._name)
PyDeadObjectError: The C++ part of the HistLimitPanel object has been deleted, attribute access no longer allowed.
ERROR:traits:Exception occurred in traits notification handler for object: <PYME.LMVis.layers.pointcloud.PointCloudRenderLayer object at 0x1835223cb0>, trait: clim, old value: [30.0, 828.0], new value: [1.0, 8000.0]
Traceback (most recent call last):
  File "/anaconda2/lib/python2.7/site-packages/traits/trait_notifiers.py", line 519, in _dispatch_change_event
    self.dispatch( handler, *args )
  File "/anaconda2/lib/python2.7/site-packages/traits/trait_notifiers.py", line 614, in dispatch
    handler( *args )
  File "/anaconda2/lib/python2.7/site-packages/traitsui/editor.py", line 366, in _update_editor
    self.update_editor()
  File "/Users/bennettrollins/python-microscopy/PYME/ui/custom_traits_editors.py", line 131, in update_editor
    self.control.SetData(self.factory.data(), *self.value)
  File "/anaconda2/lib/python2.7/site-packages/wx-3.0-osx_cocoa/wx/_core.py", line 16711, in __getattr__
    raise PyDeadObjectError(self.attrStr % self._name)
PyDeadObjectError: The C++ part of the HistLimitPanel object has been deleted, attribute access no longer allowed.

```

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Does this error actually prevent the colour update? When I've seen this (or similar) errors before they could be safely ignored.

The error can be traced back to the destruction and re-creation of the left hand panel (done to update some of the display items). Traits keeps a reference to the old display items which becomes invalid and leads to the error you have seen. I'd like to call it a traits (or traitsui) bug rather than one in PYME, but there is a chance that we could resolve it by doing some traits specific cleanup in the panel recreation code (more details to follow when I'm at my computer).

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

More details ... the root cause (in our code) is in `CreateFoldPanel()`, Line 101 in AUIFrame.py. This clears the side panel (if already present). Unfortunately the traits editors do not get cleaned up properly when we call `self.sidepanel.clear()`. This is almost certainly a bug in either  traitsui or wxpython, but we might be able to work around it by doing something like manually destroying all the panels children before we clear it (although this shouldn't be necessary). 

```
#!python

 def CreateFoldPanel(self):
        """Create a panel of folding 'drawers' on the left side of the frame.
        loops over all the functions defined in self.paneHooks and calls them
        to generate the drawers.
        """
        pinfo = self._mgr.GetPaneByName('sidePanel')
        if pinfo.IsOk(): #we already have a sidepanel, clear
            self.sidePanel.Clear()
        else:
```

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

(hopefully) fixed in 295b227.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Marking as resolved as it should be fixed in 295b227, otherwise effective workarounds  (i.e. ignoring it) exist.
